{% extends "layout.html" %}
{% set title = _('Overview') %}
{% block body %}
  <h1>Welcome to ProPhyle</h1>


  <p>ProPhyle is a phylogeny-based metagenomic classifier, whose indexing strategy is based on the bottom-up propagation of genomes’ k-mers in the tree, assembling contigs at each node and matching using a full-text search.
  </p>

  <p>
  Compared to other state-of-the-arts classifiers, ProPhyle provides several unique features:
  
  <img src="{{ pathto('_static/prophyle_overview.jpg', 1) }}"  style="float:right; width:250px" />

  <ul>
    <li><b>Low memory requirements.</b> Compared to Kraken, ProPhyle has 9x smaller memory footprint for index construction and 5x smaller footprint for querying.</li>
    <li><b>Flexibility.</b> ProPhyle is easy to use with any user-provided phylogenetic trees and reference genomes.</li>
    <li><b>Standard bioinformatics formats.</b> Newick/NHX is used for representing phylogenetic trees and SAM for reporting the assignments.</li>
    <li><b>Lossless k-mer indexing.</b> ProPhyle stores a list of all genomes containing a k-mer. It can be, therefore, accurate even with trees containing similar genomes (e.g, phylogenetic trees for a single species).</li>
    <li><b>Deterministic behavior.</b> ProPhyle is a fully deterministic classifier with a mathematically well-defined behavior.</li>
  </ul>

  <h2 style="margin-bottom: 0">Documentation</h2>

  <table class="contentstable">
    <tr>
      <td style="width:35%">
        <p class="biglink"><a class="biglink" href="{{ pathto("example") }}">Quick example</a><br/>
           <span class="linkdescr">5 minutes example</span></p>
      </td><td>
        <p class="biglink"><a class="biglink" href="{{ pathto("tutorial") }}">Tutorial</a><br/>
           <span class="linkdescr">overview of basic tasks</span></p>
      </td>
    </tr><tr>
      <td>
        {%- if hasdoc('search') %}<p class="biglink"><a class="biglink" href="{{ pathto("search") }}">Search page</a><br/>
           <span class="linkdescr">search the documentation</span></p>{%- endif %}
      </td><td>
        <p class="biglink"><a class="biglink" href="{{ pathto("contents") }}">Contents</a><br/>
           <span class="linkdescr">for a complete overview</span></p>
      </td>
    </tr><tr>
      <td>
        <p class="biglink"><a class="biglink" href="{{ pathto("reference") }}">Reference</a><br/>
           <span class="linkdescr">command-line interface and data formats</span></p>
      </td><td>
        <p class="biglink"><a class="biglink" href="https://github.com/prophyle/prophyle/releases">
          Releases
        </a><br/>
           <span class="linkdescr">release history</span></p>
      </td><td>
      </td>
    </tr>
  </table>


  <h2 style="margin-bottom: 0">Auxiliary tools</h2>

  <table class="contentstable">
    <tr>
      <td style="width:35%">
        <p class="biglink"><a class="biglink" href="http://github.com/prophyle/prophyle_index">ProPhyle Index</a><br/>
           <span class="linkdescr">BWT-index for rapid k-mer matching</span></p>
      </td>
      <td>
        <p class="biglink"><a class="biglink" href="http://github.com/prophyle/prophyle_assembler">ProPhyle Assembler</a><br/>
           <span class="linkdescr">compressing k-mers by assemblying contigs</span></p>
      </td>
    </tr>
  </table>


  <h2>Cite</h2>

  <ul>
  <li>
    K. Břinda, K. Salikhov, S. Pignotti, G. Kucherov.
    <b>ProPhyle 0.3.1.0</b>, Zenodo, 2017.
    doi:<a href="http://doi.org/10.5281/zenodo.1045429">10.5281/zenodo.1045429</a>.</li>
  </ul>


  <h2>Authors</h2>

  <ul>
  <li><a href="http://brinda.cz">Karel Brinda <<a href="">kbrinda@hsph.harvard.edu</a>></li>
  <li>Kamil Salikhov <<a href="">kamil.salikhov@univ-mlv.fr</a>></li>
  <li>Simone Pignotti <<a href="">pignottisimone@gmail.com</a>></li>
  <li><a href="http://igm.univ-mlv.fr/~koutcher/">Gregory Kucherov <<a href="">gregory.kucherov@univ-mlv.fr</a>></li>
  </ul>

{% endblock %}
